File: src/commands/memory/add.ts
import { defineCommand, option } from "@bunli/core";
import { z } from "zod";
import { createPaths } from "../../config/paths.js";
import { loadConfig } from "../../config/load-config.js";
import { FastEmbedProvider } from "../../embeddings/fastembed-provider.js";
import { runIngest } from "../../ingest/ingest-service.js";
import { readInput } from "../../ingest/read-input.js";
import { createDatabase } from "../../storage/lancedb.js";
import { renderData } from "../../output/render.js";

export default defineCommand({
  name: "add",
  description: "Add text, files, or stdin content to memory",
  options: {
    file: option(z.string().optional(), { description: "Path to a file to ingest" }),
    text: option(z.string().optional(), { description: "Direct text to ingest" }),
    stdin: option(z.coerce.boolean().default(false), { description: "Read content from stdin" }),
    title: option(z.string().optional(), { description: "Optional document title" }),
    tag: option(z.array(z.string()).default([]), { description: "Tags to attach to the document" }),
    chunkSize: option(z.coerce.number().int().positive().optional(), {
      description: "Chunk size in characters"
    }),
    chunkOverlap: option(z.coerce.number().int().min(0).optional(), {
      description: "Chunk overlap in characters"
    }),
    json: option(z.coerce.boolean().default(false), { description: "Print machine-readable output" })
  },
  handler: async ({ flags }) => {
    const paths = createPaths();
    const config = await loadConfig(paths);
    const input = await readInput({
      file: flags.file,
      text: flags.text,
      stdin: flags.stdin
    });
    const provider = new FastEmbedProvider({
      cacheDir: paths.modelsDir,
      model: config.embeddings.model,
      showDownloadProgress: false
    });
    const db = await createDatabase(paths, provider);
    const result = await runIngest({
      db,
      provider,
      config,
      input,
      title: flags.title ?? input.title,
      tags: flags.tag,
      chunkSize: flags.chunkSize,
      chunkOverlap: flags.chunkOverlap
    });
    renderData(result, flags.json);
  }
});

File: src/commands/memory/add.rs
use anyhow::{anyhow, Result};

use crate::{
    cli::MemoryAddArgs,
    input, output,
    services::{
        memory::{self, MemoryAddRequest, MemorySource},
        runtime::ServiceRuntime,
    },
};

pub async fn run(args: MemoryAddArgs) -> Result<()> {
    let runtime = ServiceRuntime::bootstrap(None, None, !args.json).await?;
    let source = match (args.file, args.text, args.stdin) {
        (Some(path), None, false) => MemorySource::File { path },
        (None, Some(text), false) => MemorySource::Text { text },
        (None, None, true) => {
            let input = input::read_input(None, None, true).await?;
            MemorySource::Stdin {
                text: input.content,
            }
        }
        _ => return Err(anyhow!("exactly one input source is required: --file, --text, or --stdin")),
    };
    let result = memory::add(
        &runtime,
        MemoryAddRequest {
            source,
            title: args.title,
            tags: args.tags,
            chunk_size: args.chunk_size,
            chunk_overlap: args.chunk_overlap,
        },
    )
    .await?;

    output::render(
        &serde_json::json!({
            "ok": result.ok,
            "documentId": result.document_id,
            "chunkCount": result.chunk_count,
            "title": result.title,
        }),
        args.json,
    )
}
